### PR TITLE
fix(dynamic_links,ios): retry handling iOS universal link on network failure

### LINKED ISSUE #33: [firebase_dynamic_links] plugin is not working in ios when app is in Back ground.
I have downloaded the example app from [https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_dynamic_links/example](url).
This example is running fine on android but on ios it is doing nothing when app is in background.
In fact onLink method was never called when ios app is in background.
HELP me please.

### LINKED ISSUE #100: [firebase_dynamic_links] Not working on app start
**Describe the bug**
When opening an iOS App from a dynamic link the initialLink is nil.

**To Reproduce**
1. Set up firebase dynamic links for a project as described in the readme.
2. Create a dynamic link (programmatically or via console)
3. Tap on link in non-browser app 

**Expected behavior**
App should open and getInitialLink() should return the correct dynamic link.

**Additional context**
From my understanding there is an issue when reacting to the incoming `UserActivity`. `onInitialLink` is correctly called but returns way earlier than `self.initialLink` is set in the completion block from `[[FIRDynamicLinks dynamicLinks]handleUniversalLink: completion: ] `.  Therefore on handling the method call `FirebaseDynamicLinks#getInitialLink` `[self getInitialLink]` returns `nil` and because of `self.flutterError` also being nil the `result` returns just with `nil`.

### PR DESCRIPTION
**Pull request by jherencia**
## Description

People have posted some issues:
- #100 
- #33 
on iOS dynamic links behavior. All of this use cases worked before iOS 12 release so I assume Firebase update and iOS 12 have changed things.

I've based my work on the release of rn-firebase 6.0 that had the same problems and were fixed so this PR takes some ideas from them.

## Related Issues

- Fixes #100 
- Fixes #33 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by jherencia** [comment]
@Salakar 

This is a clone of https://github.com/FirebaseExtended/flutterfire/pull/255 with the code updated to be mergeable.

**Comment by himsa** [comment]
Any update on this?

**Comment by b-jan** [comment]
Hello @Salakar and @jherencia 
Can we merge this one soon?

**Comment by josh-burton** [comment]
I've tested this PR and I'm unable to get a link passed through from the app store to work, unless I add a few seconds delay before checking for pending links.

File: packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m [hunk 1]
```diff
@@ -150,25 +150,61 @@ - (BOOL)checkForDynamicLink:(NSURL *)url {
...
+    retried = YES;
+    [[FIRDynamicLinks dynamicLinks] handleUniversalLink:userActivity.webpageURL
+                                             completion:completion];
+  }
+  // TODO: We could send this to Dart and maybe have a onDynamicLinkError listener but there's also
+  // a good chance the
+  // TODO: `userActivity.webpageURL` might not be for a Firebase dynamic link, which needs
+  // consideration - so we'll
+  // TODO: log this for now, logging will get picked up by Crashlytics automatically if its
+  // integrated.
```
Comment (Salakar) [inline packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m:197]:
```suggestion
  // We could send this to Dart and maybe have a onDynamicLinkError stream but there's also
  // a good chance the `userActivity.webpageURL` might not be for a Firebase dynamic link, 
  // which needs consideration - so we'll log this for now, logging will get picked up by 
  // Crashlytics automatically if its integrated.
```

Comment (Salakar) [reply packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m:197]: @jherencia please can you apply this suggestion, then I can merge. For future PRs it's helpful to allow maintainers of the repository you're submitting your PR to to push changes to your branch.

Comment (jherencia) [reply packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m]: Done

**Review by Salakar — APPROVED** [review]

**Pull request merged**